go-controller/pkg/ovn/libovsdbops/acl.go

```diff
@@ -0,0 +1,220 @@
+		existing.Action == searched.Action
+}
+
+// findACL looks up the client cache for equivalent ACLs
+func findACL(nbClient libovsdbclient.Client, acl *nbdb.ACL) (string, error) {
+	if acl.UUID != "" && !isNamedUUID(acl.UUID) {
+		return acl.UUID, nil
+	}
+
+	acls := []nbdb.ACL{}
+	err := nbClient.WhereCache(func(item *nbdb.ACL) bool {
+		return isEquivalentACL(item, acl)
+	}).List(&acls)
+	if err != nil && err != libovsdbclient.ErrNotFound {
+		return "", fmt.Errorf("can't find ACL by equivalence %+v: %v", *acl, err)
+	}
+
+	if len(acls) > 1 {
+		return "", fmt.Errorf("unexpectedly found multiple equivalent ACLs: %+v", acls)
+	}
+
+	if len(acls) == 1 {
+		return acls[0].UUID, nil
+	}
+
+	return "", nil
+}
+
+func BuildACL(name, direction string, priority int, match, action, meter, severity string, log bool, externalIds map[string]string) *nbdb.ACL {
+	uuid := buildNamedUUID()
+	name = fmt.Sprintf("%.63s", name)
+	return &nbdb.ACL{
+		UUID:        uuid,
+		Name:        &name,
+		Direction:   direction,
+		Match:       match,
+		Action:      action,
+		Priority:    priority,
+		Severity:    &severity,
+		Log:         log,
+		Meter:       &meter,
+		ExternalIDs: externalIds,
+	}
+}
+
+func CreateOrUpdateACLOps(nbClient libovsdbclient.Client, ops []libovsdb.Operation, acl *nbdb.ACL) ([]libovsdb.Operation, error) {
+	if ops == nil {
+		ops = []libovsdb.Operation{}
+	}
+
+	uuid, err := findACL(nbClient, acl)
+	if err != nil {
+		return nil, err
+	}
+
+	// If ACL already exists, update it
+	if uuid != "" {
+		acl.UUID = ""
+		op, err := nbClient.Where(&nbdb.ACL{UUID: uuid}).Update(acl)
+		if err != nil {
+			return nil, err
+		}
+		ops = append(ops, op...)
+		acl.UUID = uuid
+		return ops, nil
+	}
+
+	op, err := nbClient.Create(acl)
+	if err != nil {
+		return nil, err
+	}
+	ops = append(ops, op...)
+
+	return ops, nil
+}
+
+func CreateOrUpdateACL(nbClient libovsdbclient.Client, acl *nbdb.ACL) error {
+	ops, err := CreateOrUpdateACLOps(nbClient, nil, acl)
+	if err != nil {
+		return err
+	}
+
+	_, err = TransactAndCheck(nbClient, ops)
+	return err
+}
+
+func UpdateACLOps(nbClient libovsdbclient.Client, ops []libovsdb.Operation, acl *nbdb.ACL) ([]libovsdb.Operation, error) {
+	if ops == nil {
+		ops = []libovsdb.Operation{}
+	}
+
+	uuid, err := findACL(nbClient, acl)
+	if err != nil {
+		return nil, err
+	}
+
+	if uuid == "" {
+		return nil, fmt.Errorf("error, acl not found %+v", acl)
+	}
+
+	acl.UUID = ""
+	op, err := nbClient.Where(&nbdb.ACL{UUID: uuid}).Update(acl)
+	if err != nil {
+		return nil, err
+	}
+	ops = append(ops, op...)
+	acl.UUID = uuid
+
+	return ops, nil
+}
+
+func UpdateACL(nbClient libovsdbclient.Client, acl *nbdb.ACL) error {
+	ops, err := UpdateACLOps(nbClient, nil, acl)
+	if err != nil {
+		return err
+	}
+
+	_, err = TransactAndCheck(nbClient, ops)
+	return err
+}
+
+func UpdateACLLoggingOps(nbClient libovsdbclient.Client, ops []libovsdb.Operation, acl *nbdb.ACL) ([]libovsdb.Operation, error) {
```

Comment:
When is this useful? Are you worried about cache consistency issues?

Reply:
Are you asking about having a specific method to update log & severity?

If so, no. One reason is that I just did the functional conversion of the code we already had in place for this. Other reason is that the user does not have to know the full details of the ACL to just update the log & severity, name might be enough. Other reason is I did not have thoughts against providing any convenience methods that update specific attributes instead of the full thing. Final reason is that ovsdb might no lock the full row to just update those two columns, although this is not something I am 100% sure about.

Reply:
Seems reasonable; just curious what the reasoning was.